File: Engr-T.stark/decodeWays.cpp
class Solution {
public:
    int numDecodings(string s) {

    vector<int> dp(s.size()+2, 0);

    dp[s.size()] = 1;

    for (int i = s.size()-1; i >= 0; --i) {

        dp[i] += dp[i+1];

        if(s[i] == '1')

            dp[i] += dp[i+2];

        else if(s[i] == '2' && i+1 < s.size() && s[i+1] < '7')

            dp[i] += dp[i+2];

        else if(s[i] == '0')

            dp[i] = 0;

    }

    return dp[0];

}
        

};

File: Richie/Decode ways/decode-ways.py
class Solution:
    def numDecodings(self, s, memo = {}):
        if s in memo:
            return memo[s]
        if s and s[0] == "0":
            return 0
        if len(s)<= 1:
            return 1
        
        sn, dn = int(s[:1]), int(s[:2])
        
        result = self.numDecodings(s[1:]) 
        
        if 27>dn>0:
            result += self.numDecodings(s[2:]) 
        memo[s] = result
        return result
                     
